File: BuildTools/Protection/encryptor.py
import random
from PyQt5.QtCore import QSettings

def get_encryption_method():
    settings = QSettings('SysCaller', 'BuildTools')
    return settings.value('obfuscation/encryption_method', 1, int)

def encrypt_offset(real_offset, method=1):
    if method == 1: # Basic XOR
        key = random.randint(0x11, 0xFF)
        encrypted_offset = real_offset ^ key
        return encrypted_offset, {"key": key}
    elif method == 2: # Multi Key XOR
        key1 = random.randint(0x11, 0xFF)
        key2 = random.randint(0x11, 0xFF)
        encrypted_offset = (real_offset ^ key1) ^ key2
        return encrypted_offset, {"key1": key1, "key2": key2}
    elif method == 3: # Add + XOR combination
        add_val = random.randint(0x100, 0xFFF)
        xor_key = random.randint(0x11, 0xFF)
        encrypted_offset = (real_offset + add_val) ^ xor_key
        return encrypted_offset, {"add_val": add_val, "xor_key": xor_key}
    elif method == 4: # Enhanced XOR
        xor_key = random.randint(0x1000, 0xFFFF)
        encrypted_offset = real_offset ^ xor_key
        return encrypted_offset, {"xor_key": xor_key}
    elif method == 5: # Offset Shifting
        mask = random.randint(0x100, 0xFFF)
        encrypted_offset = (real_offset + mask) & 0xFFFFFFFF
        return encrypted_offset, {"mask": mask}
    key = random.randint(0x11, 0xFF) # default to basic XOR if invalid
    encrypted_offset = real_offset ^ key
    return encrypted_offset, {"key": key}

def generate_decryption_sequence(offset_name, encryption_data, method=1):
    if method == 1: # Basic XOR
        return [
            f"    mov eax, dword ptr [{offset_name}]\n",
            f"    mov ebx, 0{encryption_data['key']:X}h\n",
            f"    xor eax, ebx\n"
        ]
    elif method == 2: # Multi Key XOR
        return [
            f"    mov eax, dword ptr [{offset_name}]\n",
            f"    mov ebx, 0{encryption_data['key1']:X}h\n",
            f"    xor eax, ebx\n",
            f"    mov ebx, 0{encryption_data['key2']:X}h\n",
            f"    xor eax, ebx\n"
        ]
    elif method == 3: # Add + XOR combination
        return [
            f"    mov eax, dword ptr [{offset_name}]\n",
            f"    mov ebx, 0{encryption_data['xor_key']:X}h\n",
            f"    xor eax, ebx\n",
            f"    sub eax, 0{encryption_data['add_val']:X}h\n"
        ]
    elif method == 4: # Enhanced XOR
        xor_key = encryption_data['xor_key']
        return [
            f"    mov eax, dword ptr [{offset_name}]\n",
            f"    mov ebx, 0{xor_key:X}h\n",
            f"    xor eax, ebx\n"
        ]
    elif method == 5: # Offset Shifting
        mask = encryption_data['mask']
        return [
            f"    mov eax, dword ptr [{offset_name}]\n",
            f"    sub eax, 0{mask:X}h\n"
        ]
    return [ # default to basic XOR if invalid
        f"    mov eax, dword ptr [{offset_name}]\n",
        f"    mov ebx, 0{encryption_data['key']:X}h\n",
        f"    xor eax, ebx\n"
    ] 

File: Bind/src/Core/Obfuscation/Encryption/Encryptor.cpp
#include "include/Core/Obfuscation/Encryption/Encryptor.h"
#include <QRandomGenerator>
#include <QDebug>

Encryptor::Encryptor(QSettings* settings) : settings(settings) {
}

void Encryptor::setSettings(QSettings* settings) {
    this->settings = settings;
}

int Encryptor::getEncryptionMethod() {
    if (!settings) {
        return 1;
    }
    return settings->value("obfuscation/encryption_method", 1).toInt();
}

QPair<int, QMap<QString, QVariant>> Encryptor::encryptOffset(int realOffset, int method) {
    if (method == -1) {
        method = getEncryptionMethod();
    }
    QMap<QString, QVariant> encryptionData;
    int encryptedOffset;
    switch (method) {
        case 1: { // basic xor
            int key = getRandomInt(0x11, 0xFF);
            encryptedOffset = realOffset ^ key;
            encryptionData["key"] = key;
            break;
        }
        case 2: { // multi key xor
            int key1 = getRandomInt(0x11, 0xFF);
            int key2 = getRandomInt(0x11, 0xFF);
            encryptedOffset = (realOffset ^ key1) ^ key2;
            encryptionData["key1"] = key1;
            encryptionData["key2"] = key2;
            break;
        }
        case 3: { // add + xor combo
            int addVal = getRandomInt(0x100, 0xFFF);
            int xorKey = getRandomInt(0x11, 0xFF);
            encryptedOffset = (realOffset + addVal) ^ xorKey;
            encryptionData["add_val"] = addVal;
            encryptionData["xor_key"] = xorKey;
            break;
        }
        case 4: { // enhanced xor
            int xorKey = getRandomInt(0x1000, 0xFFFF);
            encryptedOffset = realOffset ^ xorKey;
            encryptionData["xor_key"] = xorKey;
            break;
        }
        case 5: { // offset shifting
            int mask = getRandomInt(0x100, 0xFFF);
            encryptedOffset = (realOffset + mask) & 0xFFFFFFFF;
            encryptionData["mask"] = mask;
            break;
        }
        default: { // default to basic xor
            int key = getRandomInt(0x11, 0xFF);
            encryptedOffset = realOffset ^ key;
            encryptionData["key"] = key;
            break;
        }
    }
    return qMakePair(encryptedOffset, encryptionData);
}

QStringList Encryptor::generateDecryptionSequence(const QString& offsetName, const QMap<QString, QVariant>& encryptionData, int method) {
    if (method == -1) {
        method = getEncryptionMethod();
    }
    QStringList sequence;
    switch (method) {
        case 1: { // basic xor
            int key = encryptionData["key"].toInt();
            sequence << QString("    mov eax, dword ptr [%1]\n").arg(offsetName);
            sequence << QString("    mov ebx, 0%1h\n").arg(key, 0, 16);
            sequence << "    xor eax, ebx\n";
            break;
        }
        case 2: { // multi key xor
            int key1 = encryptionData["key1"].toInt();
            int key2 = encryptionData["key2"].toInt();
            sequence << QString("    mov eax, dword ptr [%1]\n").arg(offsetName);
            sequence << QString("    mov ebx, 0%1h\n").arg(key1, 0, 16);
            sequence << "    xor eax, ebx\n";
            sequence << QString("    mov ebx, 0%1h\n").arg(key2, 0, 16);
            sequence << "    xor eax, ebx\n";
            break;
        }
        case 3: { // add + xor combo
            int xorKey = encryptionData["xor_key"].toInt();
            int addVal = encryptionData["add_val"].toInt();
            sequence << QString("    mov eax, dword ptr [%1]\n").arg(offsetName);
            sequence << QString("    mov ebx, 0%1h\n").arg(xorKey, 0, 16);
            sequence << "    xor eax, ebx\n";
            sequence << QString("    sub eax, 0%1h\n").arg(addVal, 0, 16);
            break;
        }
        case 4: { // enhanced xor
            int xorKey = encryptionData["xor_key"].toInt();
            sequence << QString("    mov eax, dword ptr [%1]\n").arg(offsetName);
            sequence << QString("    mov ebx, 0%1h\n").arg(xorKey, 0, 16);
            sequence << "    xor eax, ebx\n";
            break;
        }
        case 5: { // offset shifting
            int mask = encryptionData["mask"].toInt();
            sequence << QString("    mov eax, dword ptr [%1]\n").arg(offsetName);
            sequence << QString("    sub eax, 0%1h\n").arg(mask, 0, 16);
            break;
        }
        default: { // default to basic xor
            int key = encryptionData["key"].toInt();
            sequence << QString("    mov eax, dword ptr [%1]\n").arg(offsetName);
            sequence << QString("    mov ebx, 0%1h\n").arg(key, 0, 16);
            sequence << "    xor eax, ebx\n";
            break;
        }
    }
    return sequence;
}

int Encryptor::getRandomInt(int min, int max) {
    return QRandomGenerator::global()->bounded(min, max + 1);
} 